nym-vpn-lib/src/credentials/check.rs

```diff
@@ -0,0 +1,107 @@
+use std::{fs, path::PathBuf};
+
+use nym_bandwidth_controller::{BandwidthController, PreparedCredential, RetrievedCredential};
+use nym_credentials::{
+    coconut::bandwidth::{
+        bandwidth_credential_params, issued::BandwidthCredentialIssuedDataVariant,
+    },
+    obtain_aggregate_verification_key, IssuedBandwidthCredential,
+};
+
+use tracing::{debug, info};
+
+use super::{
+    helpers::{get_coconut_api_clients, get_credentials_store, get_nyxd_client, CoconutClients},
+    CredentialError,
+};
+
+pub async fn check_raw_credential(raw_credential: Vec<u8>) -> Result<(), CredentialError> {
+    let version = None;
+    let credential = IssuedBandwidthCredential::try_unpack(&raw_credential, version)?;
+
+    // Check expiry
+    match credential.variant_data() {
+        BandwidthCredentialIssuedDataVariant::Voucher(_) => {
+            debug!("credential is a bandwidth voucher");
+        }
+        BandwidthCredentialIssuedDataVariant::FreePass(freepass_info) => {
+            debug!("credential is a free pass");
+            if freepass_info.expired() {
+                return Err(CredentialError::FreepassExpired {
+                    expiry_date: freepass_info.expiry_date().to_string(),
+                });
+            }
+        }
+    }
+
+    // TODO: verify?
+
+    Ok(())
+}
+
+pub async fn check_credential_base58(credential: &str) -> Result<(), CredentialError> {
+    let raw_credential = bs58::decode(credential)
+        .into_vec()
+        .map_err(|err| CredentialError::FailedToDecodeBase58Credential { source: err })?;
+    check_raw_credential(raw_credential).await
+}
+
+pub async fn check_credential_file(credential_file: PathBuf) -> Result<(), CredentialError> {
+    let raw_credential = fs::read(credential_file)?;
+    check_raw_credential(raw_credential).await
+}
+
+pub async fn check_imported_credential(
```

Comment:
this doesn't seem to be used? 
plus it suffers from the same issue we had before -> it verifies the "oldest" available credential rather than the one you just imported

Reply:
It's called before connect